File: yukino/teams/teams_test.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package teams

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
)

func TestFileMailBoxRoundTrip(t *testing.T) {
	dir := t.TempDir()
	mb := NewFileMailBox(dir)

	if err := mb.Send("alice", FileMailMessage{From: "bob", Text: "hi"}); err != nil {
		t.Fatalf("send: %v", err)
	}
	if err := mb.Send("alice", FileMailMessage{From: "carol", Text: "hello"}); err != nil {
		t.Fatalf("send: %v", err)
	}

	unread := mb.ReadUnread("alice")
	if len(unread) != 2 {
		t.Fatalf("expected 2 unread, got %d", len(unread))
	}

	if err := mb.MarkAllRead("alice"); err != nil {
		t.Fatalf("mark read: %v", err)
	}
	unread2 := mb.ReadUnread("alice")
	if len(unread2) != 0 {
		t.Errorf("expected 0 unread after MarkAllRead, got %d", len(unread2))
	}
}

func TestFileMailBoxConcurrentSends(t *testing.T) {
	dir := t.TempDir()
	mb := NewFileMailBox(dir)

	const n = 20
	var wg sync.WaitGroup
	sendErrs := make(chan error, n)
	for i := range n {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			if err := mb.Send("dest", FileMailMessage{From: "sender", Text: "msg"}); err != nil {
				sendErrs <- err
			}
		}(i)
	}
	wg.Wait()
	close(sendErrs)
	// Send failures directly cause missing messages in the inbox; surface
	// errors first to avoid only seeing a count mismatch.
	for err := range sendErrs {
		t.Errorf("send failed: %v", err)
	}

	got := mb.ReadUnread("dest")
	if len(got) != n {
		t.Errorf("expected %d messages after concurrent sends, got %d", n, len(got))
	}
}

func TestTeamManagerCRUD(t *testing.T) {
	// Each test case uses an isolated teams directory to prevent messages from
	// accumulating in the same inbox across runs.
	dir := t.TempDir()
	tm := NewTeamManager(dir, dir)

	team := tm.CreateTeam("alpha")
	if team == nil {
		t.Fatal("CreateTeam returned nil")
	}
	if got := tm.GetTeam("alpha"); got != team {
		t.Errorf("Get should return same team instance")
	}
	if names := tm.ListTeams(); len(names) != 1 || names[0] != "alpha" {
		t.Errorf("ListTeams = %v, want [alpha]", names)
	}
	tm.DeleteTeam("alpha")
	if got := tm.GetTeam("alpha"); got != nil {
		t.Error("DeleteTeam did not remove team")
	}
}

// TestSendMessageToolRoutesToLead pins the fix for the bug where a
// teammate calling SendMessage(to="lead", ...) saw "recipient 'lead' not
// found in any team" because the lead is never registered as a Member.
// The tool must recognize LeadName and route via the sender's team
// mailbox so the lead can read the reply on its next sweep.
func TestSendMessageToolRoutesToLead(t *testing.T) {
	// Each test case uses an isolated teams directory to prevent messages from
	// accumulating in the same inbox across runs.
	dir := t.TempDir()
	tm := NewTeamManager(dir, dir)
	team := tm.CreateTeam("demo")
	team.AddMember("alice")

	tool := &SendMessageTool{TeamMgr: tm, SenderName: "alice"}
	res := tool.Execute(context.Background(), map[string]any{
		"to":      LeadName,
		"content": "here is the README summary",
	})
	if res.IsError {
		t.Fatalf("SendMessage to lead errored: %s", res.Output)
	}
	if !strings.Contains(res.Output, LeadName) {
		t.Errorf("expected confirmation mentioning %q, got %q", LeadName, res.Output)
	}

	msgs := team.MailBox.ReadUnread(LeadName)
	if len(msgs) != 1 {
		t.Fatalf("expected 1 message in lead inbox, got %d", len(msgs))
	}
	if msgs[0].From != "alice" || msgs[0].Text != "here is the README summary" {
		t.Errorf("unexpected message: %+v", msgs[0])
	}
}

// TestSendMessageToolUnknownSenderFallsBackToFirstTeam pins the TS senderTeam
// semantics (tools.ts:177-185): a sender not on any roster falls back to the
// first team — only when NO team exists does the tool surface "No active team
// found for this sender."
func TestSendMessageToolUnknownSenderFallsBackToFirstTeam(t *testing.T) {
	dir := t.TempDir()
	tm := NewTeamManager(dir, dir)
	team := tm.CreateTeam("demo") // no members added

	tool := &SendMessageTool{TeamMgr: tm, SenderName: "ghost"}
	res := tool.Execute(context.Background(), map[string]any{
		"to":      LeadName,
		"content": "anyone?",
	})
	if res.IsError {
		t.Fatalf("TS falls back to the first team for an unknown sender: %s", res.Output)
	}
	msgs := team.MailBox.ReadUnread(LeadName)
	if len(msgs) != 1 {
		t.Fatalf("expected 1 message in lead inbox, got %d", len(msgs))
	}

	// With no team at all the send must fail loudly.
	empty := NewTeamManager(dir, dir)
	res = (&SendMessageTool{TeamMgr: empty, SenderName: "ghost"}).Execute(
		context.Background(), map[string]any{"to": LeadName, "content": "anyone?"})
	if !res.IsError || res.Output != "No active team found for this sender." {
		t.Fatalf("expected the no-team error, got: %+v", res)
	}
}

var _ = filepath.Join
var _ = os.Getwd

File: apps/yukino/tests/teams.test.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { describe, it, expect, beforeEach, afterEach } from "vitest";

import { TeamManager } from "@/teams/index.js";
import {
  createProgress,
  recordTokens,
  recordToolResult,
  recordToolStart,
  recordTurnComplete,
} from "@/teams/progress.js";
import { listTeamNames } from "@/teams/team-file.js";
import {
  TeamCreateTool,
  SpawnTeammateTool,
  SendMessageTool,
  ListTeamsTool,
} from "@/teams/tools.js";

// The teams directory lives at <home>/.yukino/teams, so the tests redirect the
// entire home directory to a temp dir to avoid leaving residue in the real
// ~/.yukino/teams. os.homedir() reads USERPROFILE on Windows and HOME on other
// platforms, so set both.
let realHome: string | undefined;
let realUserProfile: string | undefined;
beforeEach(() => {
  realHome = process.env.HOME;
  realUserProfile = process.env.USERPROFILE;
  const tmp = mkdtempSync(join(tmpdir(), "yukino-home-"));
  process.env.HOME = tmp;
  process.env.USERPROFILE = tmp;
});
afterEach(() => {
  if (realHome === undefined) {
    delete process.env.HOME;
  } else {
    process.env.HOME = realHome;
  }
  if (realUserProfile === undefined) {
    delete process.env.USERPROFILE;
  } else {
    process.env.USERPROFILE = realUserProfile;
  }
});
const wait = (ms: number) => new Promise((r) => setTimeout(r, ms));
const workDir = () => mkdtempSync(join(tmpdir(), "yukino-team-"));

describe("teammate progress", () => {
  it("tracks active tools, turns, and cumulative tokens", () => {
    const progress = createProgress();

    recordToolStart(progress, "read", "ReadFile", { file_path: "a.ts" });
    recordToolStart(progress, "bash", "Bash", { command: "pwd" });
    recordTokens(progress, 100, 20);
    recordTokens(progress, 200, 30);

    expect(progress.activeTools.at(-1)?.toolName).toBe("Bash");
    expect(progress.tokenCount).toBe(350);

    recordToolResult(progress, "bash");
    expect(progress.activeTools.at(-1)?.toolName).toBe("ReadFile");

    recordTurnComplete(progress);
    expect(progress.turnCount).toBe(1);
    expect(progress.activeTools).toEqual([]);
  });
});

describe("teams orchestration", () => {
  it("spawnTeammate runs the task and posts its result to the lead mailbox", async () => {
    const mgr = new TeamManager(workDir());
    const team = mgr.create("squad");
    team.spawnTeammate(
      "scout",
      "find X",
      (task) => Promise.resolve(`did: ${task}`),
      undefined,
      undefined,
      "agent-tool-call",
    );
    expect(team.getMember("scout")?.uiState?.originToolCallId).toBe(
      "agent-tool-call",
    );

    await wait(200);
    const drained = mgr.drainLeads();
    // The teammate sends an [idle] notification with its name after finishing
    expect(
      drained.some((d) => d.includes("scout") && d.includes("[idle]")),
    ).toBe(true);
    // Drained messages are consumed.
    expect(mgr.drainLeads()).toEqual([]);
  });

  it("a failing teammate reports the error to the lead", async () => {
    const mgr = new TeamManager(workDir());
    // eslint-disable-next-line @typescript-eslint/require-await
    mgr.create("squad").spawnTeammate("flaky", "boom", async () => {
      throw new Error("kaboom");
    });
    await wait(200);
    expect(mgr.drainLeads().some((d) => d.includes("failed"))).toBe(true);
  });

  it("TaskStop aborts an active in-process teammate and waits for it to settle", async () => {
    const mgr = new TeamManager(workDir());
    let resolveStarted!: (signal: AbortSignal) => void;
    let cancelled = false;
    const started = new Promise<AbortSignal>((resolve) => {
      resolveStarted = resolve;
    });

    const team = mgr.create("squad");
    team.spawnTeammate(
      "scout",
      "long task",
      async (_task, _onEvent, signal) => {
        if (!signal) {
          throw new Error("missing teammate abort signal");
        }
        resolveStarted(signal);
        await new Promise<void>((resolve) => {
          if (signal.aborted) {
            resolve();
            return;
          }
          signal.addEventListener(
            "abort",
            () => {
              cancelled = true;
              resolve();
            },
            { once: true },
          );
        });
        return "[Interrupted]";
      },
    );

    const signal = await started;
    await team.stopMember("scout");

    expect(signal.aborted).toBe(true);
    expect(cancelled).toBe(true);
    expect(team.getMember("scout")?.active).toBe(false);
    expect(team.getMember("scout")?.uiState?.status).toBe("stopped");
    expect(
      mgr.drainLeads().some((message) => message.includes("reason: stopped")),
    ).toBe(true);
  });

  it("cancels every teammate before waiting for shutdown", async () => {
    const team = new TeamManager(workDir()).create("squad");
    const cancelled: string[] = [];
    let finishFirst!: () => void;
    let finishSecond!: () => void;
    const first = team.addMember("first");
    first.active = true;
    first.cancel = () => {
      cancelled.push("first");
    };
    first.done = new Promise<void>((resolve) => {
      finishFirst = resolve;
    });
    const second = team.addMember("second");
    second.active = true;
    second.cancel = () => {
      cancelled.push("second");
    };
    second.done = new Promise<void>((resolve) => {
      finishSecond = resolve;
    });

    const stopping = team.stopAll();
    expect(cancelled).toEqual(["first", "second"]);
    finishFirst();
    finishSecond();
    await stopping;
  });

  it("coordination tools create, spawn, message, and list", async () => {
    const mgr = new TeamManager(workDir());

    expect(
      (
        await new TeamCreateTool(mgr).execute(
          {
            workDir: workDir(),
          },
          { team_name: "t1" },
        )
      ).output,
    ).toContain("created");

    // eslint-disable-next-line @typescript-eslint/require-await
    const spawn = new SpawnTeammateTool(mgr, async (task) => `done:${task}`);
    const r = await spawn.execute(
      {
        workDir: workDir(),
      },
      { team: "t1", name: "w1", task: "task A" },
    );
    expect(r.isError).toBe(false);
    await wait(200);
    expect(
      mgr.drainLeads().some((d) => d.includes("w1") && d.includes("[idle]")),
    ).toBe(true);

    // SendMessage to an existing member lands in that member's mailbox.
    const send = await new SendMessageTool(mgr).execute(
      {
        workDir: workDir(),
      },
      { to: "w1", content: "hi" },
    );
    expect(send.isError).toBe(false);
    expect(
      mgr
        .get("t1")
        ?.getMember("w1")
        ?.mailbox.receiveSync()
        .map((m) => m.text),
    ).toContain("hi");

    const list = await new ListTeamsTool(mgr).execute();
    expect(list.output).toContain("t1");
    expect(list.output).toContain("w1");
  });

  it("SendMessage delivers plain text from a teammate to the lead mailbox", async () => {
    const mgr = new TeamManager(workDir());
    mgr.create("t2").addMember("w2");

    // The lead is not a registered member, so the plain-text path must route
    // to the dedicated lead mailbox instead of throwing "Member 'lead' not found".
    const send = await new SendMessageTool(mgr, "w2").execute(
      { workDir: workDir() },
      { to: "lead", content: "findings: X confirmed" },
    );
    expect(send.isError).toBe(false);
    expect(send.output).toContain("lead");

    // The report reaches the lead as a task notification in from=X: text form.
    expect(
      mgr
        .drainLeads()
        .some((d) => d.includes("from=w2: findings: X confirmed")),
    ).toBe(true);
  });

  it("TeamCreate sweeps other teams so at most one exists", async () => {
    const mgr = new TeamManager(workDir());

    // A live team with a spawned teammate.
    await new TeamCreateTool(mgr).execute(
      { workDir: workDir() },
      { team_name: "old" },
    );
    const spawn = new SpawnTeammateTool(
      mgr,
      // eslint-disable-next-line @typescript-eslint/require-await
      async (task) => `done:${task}`,
    );
    await spawn.execute(
      { workDir: workDir() },
      { team: "old", name: "w1", task: "task A" },
    );
    await wait(200);

    // A disk-only leftover from a previous session, unknown to this manager.
    new TeamManager(workDir()).create("stale");
    expect(listTeamNames().sort()).toEqual(["old", "stale"]);

    const result = await new TeamCreateTool(mgr).execute(
      { workDir: workDir() },
      { team_name: "fresh" },
    );

    expect(result.isError).toBe(false);
    expect(result.output).toContain("fresh");
    // Exactly one team remains — in memory and on disk.
    expect(mgr.list().map((team) => team.name)).toEqual(["fresh"]);
    expect(listTeamNames()).toEqual(["fresh"]);
  });

  it("validates required args", async () => {
    const mgr = new TeamManager(workDir());
    expect(
      (
        await new TeamCreateTool(mgr).execute(
          {
            workDir: workDir(),
          },
          {},
        )
      ).isError,
    ).toBe(true);
    expect(
      (
        await new SpawnTeammateTool(
          mgr,
          // eslint-disable-next-line @typescript-eslint/require-await
          async () => "x",
        ).execute(
          {
            workDir: workDir(),
          },
          { team: "t" },
        )
      ).isError,
    ).toBe(true);
    expect(
      (
        await new SendMessageTool(mgr).execute(
          {
            workDir: workDir(),
          },
          { to: "a", content: "m" },
        )
      ).isError,
    ).toBe(true);
  });
});
